src/libcommonserver/io/iohelper.cpp

```diff
@@ -507,4 +535,104 @@ bool IoHelper::createSymlink(const SyncPath &targetPath, const SyncPath &path, I
 
     return ioError == IoErrorSuccess;
 }
+
+
+// DirectoryIterator
+
+DirectoryIterator::DirectoryIterator(const SyncPath &directoryPath, bool recursive, IoError &ioError, DirectoryOptions option)
+    : _recursive(recursive), _directoryPath(directoryPath) {
+    std::error_code ec;
+
+    _skipPermissionDenied = (option == DirectoryOptions::skip_permission_denied);
+    _dirIterator = std::filesystem::begin(std::filesystem::recursive_directory_iterator(directoryPath, option, ec));
+    ioError = IoHelper::stdError2ioError(ec);
+}
+
+
+bool DirectoryIterator::next(DirectoryEntry &nextEntry, IoError &ioError) {
+    std::error_code ec;
+    if (_invalid) {
+        ioError = IoErrorInvalidDirectoryIterator;
+        return IoHelper::_isExpectedError(ioError);
+    }
+
+    if (_directoryPath == "") {
+        ioError = IoErrorInvalidArgument;
+        return IoHelper::_isExpectedError(ioError);
+    }
+
+    if (_dirIterator == std::filesystem::end(std::filesystem::recursive_directory_iterator(_directoryPath, ec))) {
+        ioError = IoErrorEndOfDirectory;
+        return IoHelper::_isExpectedError(ioError);
+    }
+
+    if (!_recursive) {
+        disableRecursionPending();
+    }
+
+    if (!_firstElement) {
+        _dirIterator.increment(ec);
+        ioError = IoHelper::stdError2ioError(ec);
+
+        if (ioError == IoErrorNoSuchFileOrDirectory) {
+            ioError = IoErrorInvalidDirectoryIterator;
+            _invalid = true;
+            return IoHelper::_isExpectedError(ioError);
+        }
+
+        if (ioError != IoErrorSuccess) {
+            return IoHelper::_isExpectedError(ioError);
+        }
+
+    } else {
+        _firstElement = false;
+    }
+
+    if (_dirIterator != std::filesystem::end(std::filesystem::recursive_directory_iterator(_directoryPath, ec))) {
+        ioError = IoHelper::stdError2ioError(ec);
+        if (ioError == IoErrorNoSuchFileOrDirectory) {
+            ioError = IoErrorInvalidDirectoryIterator;
```

Comment:
Is it normal that you change the value of 'oError' here? 'ioError' should refelct the error returned by the stdlib

Reply:
`If a file or a directory is deleted or added to the directory tree after the directory iterator has been created, it is unspecified whether the change would be observed through the iterator.` (https://en.cppreference.com/w/cpp/filesystem/directory_iterator)

As there is no defined behaviour (i.e: the error returned by stdlib might not be consistent between OS), It seems safer to invalidate the iterator as soon as we detect a change in the directory tree? 

--- Observed behaviour on windows ---
* A file or sub directory is deleted/renamed:
    *  If we already iterate over it: The iterator still valid.
     * Else: When the iterator reach the deleted object, _dirIterator.increment() will return IoErrorNoSuchFileOrDirectory and then IoErrorUnknown to all call. -> This is why I set ioError to IoErrorInvalidDirectoryIterator to get a consistent behaviour.

* A file or sub directory is added:
    *  it will be ignored by the current iterator.

Reply:
Discussed together: it would be good to make it more genric. So if an error is detected, invalid the iterator anyway and keep the 'ioError' as it was returned by 'IoHelper::stdError2ioError(ec)' to give a hint of waht happend to the caller.
This new behavior will be verified in unit test, especially if the access rights to a folder is removed while iterating in it.